packages/yew-router/src/service.rs

```diff
@@ -1,182 +1,79 @@
-                STATE::default()
-            });
-
-            // Can't use the existing location, because this is a callback, and can't move it in
-            // here.
-            let location: Location = web_sys::window().unwrap().location();
-            let route: String = Self::get_route_from_location(&location);
-
-            callback.emit(Route { route, state })
-        };
-
-        self.event_listener = Some(EventListener::new(
-            web_sys::window().unwrap().as_ref(),
-            "popstate",
-            move |event| {
-                let event: PopStateEvent = event.clone().dyn_into().unwrap();
-                cb(event)
-            },
-        ));
-    }
-
-    /// Sets the browser's url bar to contain the provided route,
-    /// and creates a history entry that can be navigated via the forward and back buttons.
-    ///
-    /// The route should be a relative path that starts with a `/`.
-    pub fn set_route(&mut self, route: &str, state: STATE) {
-        let state_string: String = serde_json::to_string(&state).unwrap_or_else(|_| {
-            log::error!("Could not serialize state string");
-            "".to_string()
-        });
-        let _ = self
-            .history
-            .push_state_with_url(&Value::from_str(&state_string), "", Some(route));
-    }
-
-    /// Replaces the route with another one removing the most recent history event and
-    /// creating another history event in its place.
-    pub fn replace_route(&mut self, route: &str, state: STATE) {
-        let state_string: String = serde_json::to_string(&state).unwrap_or_else(|_| {
-            log::error!("Could not serialize state string");
-            "".to_string()
-        });
-        let _ =
-            self.history
-                .replace_state_with_url(&Value::from_str(&state_string), "", Some(route));
-    }
-
-    /// Gets the concatenated path, query, and fragment.
-    pub fn get_route(&self) -> Route<STATE> {
-        let route_string = Self::get_route_from_location(&self.location);
-        let state: STATE = get_state_string(&self.history)
-            .or_else(|| {
-                log::trace!("History state is empty");
-                None
-            })
-            .and_then(|state_string| -> Option<STATE> {
-                serde_json::from_str(&state_string)
-                    .ok()
-                    .or_else(|| {
-                        log::error!("Could not deserialize state string");
-                        None
-                    })
-                    .and_then(std::convert::identity) // flatten
-            })
-            .unwrap_or_default();
-        Route {
-            route: route_string,
-            state,
-        }
-    }
+    let url = web_sys::Url::new(&yew::utils::document().url().unwrap()).unwrap();
+    let parsed =
+        serde_urlencoded::from_str(&url.search().strip_prefix("?").as_deref().unwrap_or(""))?;
+    Ok(parsed)
```

Comment:
We can save a few js calls by using the [`Location`](https://rustwasm.github.io/wasm-bindgen/api/web_sys/struct.Location.html#method.search) object directly instead of constructing a `URL` object first: 
```suggestion
    let query = yew::utils::document().location().unwrap().search().unwrap();
    serde_urlencoded::from_str(query.strip_prefix("?").unwrap_or(""))
```